Index: Makefile.am
===================================================================
--- Makefile.am	(revision 13221)
+++ Makefile.am	(working copy)
@@ -7,2 +7,3 @@
		memcheck \
+		memview \
		cachegrind \
Index: configure.ac
===================================================================
--- configure.ac	(revision 13715)
+++ configure.ac	(working copy)
@@ -5173,4 +5173,6 @@
    massif/tests/Makefile
    massif/ms_print
+   memview/Makefile
+   memview/tests/Makefile
    dhat/Makefile
    dhat/tests/Makefile
